### DISCUSSION TITLE
Adding custom top-panel to theia built application.

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-theia/theia#11172] [Q&A] [answered]
How to add a custom panel to the top or below the top-panel(where menu options are available) ?

I have gone through some of the community forum discussions where I found out something called application-shell.ts file, which gives the freedom for manipulating the top level widgets of the theia built application.

The problem is , I don't have the access to the main file where the implementations of those top-level widgets are done.

Any suggestions would be of great help to me.

For the information I'm attaching the snap of the application(not the actual one) where I need the implementation.
<img width="789" alt="required_Implementaion" src="https://user-images.githubusercontent.com/57217189/169016623-d0cc51c1-d212-4f03-ae3e-4477a0227c47.PNG">

**Comment by msujew** [comment, marked as answer]
Hi @Abhay0294,

there are quite a few ways to accomplish what you're looking for. For example, Theia has a [toolbar](https://github.com/eclipse-theia/theia/blob/master/packages/toolbar/src/browser/application-shell-with-toolbar-override.ts), which inserts itself between the main area and the top panel. Also, Arduino have done [something similar](https://github.com/arduino/arduino-ide/tree/main/arduino-ide-extension/src/browser/toolbar) to this.

What do you mean by not "having the access to the main file"? Everything in Theia is freely overridable by downstream apps, so there shouldn't be an issue with missing access.

**Reply by Abhay0294** [reply]
Thanks @msujew, for the suggestion.
I'm new to eclipse theia so I was just playing around for its implementation.
Still I have a few question:
1. In which file can I code its(top panel creation) implementation, because I am having only widgets files and if I try for the implementation in a specific widget, will that implementation be not limited to that widget only. I want it to be centrally visible irrespective of the widgets.
(If possible please help with the code snippets)

**Reply by msujew** [reply]
> I want it to be centrally visible irrespective of the widgets.

Are you trying to add (or maybe even replace) widgets to the existing top panel? As far as I understood your initial question, it was about adding a second panel right under the existing top panel. For that you can reuse whatever the [toolbar](https://github.com/eclipse-theia/theia/blob/master/packages/toolbar/src/browser/application-shell-with-toolbar-override.ts) does, which should be exactly what you are looking for.

> In which file can I code its(top panel creation) implementation

When overriding the `ApplicationShell` class (like in the toolbar example above), you can also freely override the `createTopPanel` method, which is in charge of creating the top panel. Keep in mind that other contributions still might want to add new widgets to the top panel. For example, the [`BrowserMenuBarContribution`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/menu/browser-menu-plugin.ts#L369) will still add the menu bar to the top panel, even if you override the creation of the top panel itself.

https://github.com/eclipse-theia/theia/blob/70466bb4b7283187b019defec095f8dea42d47c6/packages/core/src/browser/shell/application-shell.ts#L532-L537

**Reply by Abhay0294** [reply]
Thanks a lot @msujew, I'll try with this approach.
